### PR TITLE
Thrown projectiles display item methods API

### PR DESCRIPTION
**Pull request by ghost**
Solves #2538.

This is my last and probably the best attempt (as it is the smallest API I could come up with and is pretty easy to maintain) to add API that will allow developers to change thrown projectiles' display item without having to access NMS.

Tested with the following code:
```java
    @EventHandler
    public void onProjectileLaunch(ProjectileLaunchEvent event) {
        Projectile projectile = event.getEntity();
        if (projectile instanceof Snowball) {
            ((Snowball) projectile).setDisplayItem(new ItemStack(Material.ENDER_PEARL));
        } else if (projectile instanceof EnderPearl) {
            ((EnderPearl) projectile).setDisplayItem(new ItemStack(Material.SNOWBALL));
        } else if (projectile instanceof ThrownProjectile) {
            ((ThrownProjectile) projectile).setDisplayItem(new ItemStack(Material.BLUE_WOOL));
        }
    }
```

### CODE REVIEW

File: Spigot-API-Patches/0186-Thrown-projectiles-display-item-methods.patch [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+diff --git a/src/main/java/org/bukkit/entity/ThrownExpBottle.java b/src/main/java/org/bukkit/entity/ThrownExpBottle.java
+index 671282ed..2a6e0c7b 100644
+--- a/src/main/java/org/bukkit/entity/ThrownExpBottle.java
++++ b/src/main/java/org/bukkit/entity/ThrownExpBottle.java
+@@ -3,6 +3,6 @@ package org.bukkit.entity;
+ /**
+  * Represents a thrown Experience bottle.
+  */
+-public interface ThrownExpBottle extends Projectile {
++public interface ThrownExpBottle extends Projectile { // Paper
```
Comment (BillyGalbreath) [inline Spigot-API-Patches/0186-Thrown-projectiles-display-item-methods.patch]: Did you miss this one, or intentional?

Comment (ghost) [reply Spigot-API-Patches/0186-Thrown-projectiles-display-item-methods.patch]: I missed this one, will commit a fix in a moment.

**Comment by silverwolfg11** [comment]
Is it possible to also apply this to Tridents as well? Since tridents are well projectiles and thrown, or is that out of the scope of this PR?

**Comment by ghost** [comment]
> Is it possible to also apply this to Tridents as well? Since tridents are well projectiles and thrown, or is that out of the scope of this PR?

It's not possible for now. I only implemented this API to projectiles that extend EntityProjectileThrowable (a class that allows changing the display item), and thus this API does not target arrows, potions, or trident.

I tried making the API work with potions but they just do not seem to care about completely custom display item no matter what I do. Trying to make it work with trident/arrows will possibly break API, requires mapping changes (as there are some methods with the same name and return type but they do completely something else and some requires different return type but that will most likely force us to make CraftBukkit and/or API changes), and will most likely not work anyway.

**Review by Proximyst — APPROVED** [review]
Rebase (& new PR) needed, but LGTM.

**Comment by MiniDigger** [comment]
thanks (and sorry for not getting to this sooner.......), but upstream got a api for this
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/6af4c0b2cc5a2765b8c413f5a962fdf5b4d14f84

**Pull request closed without merging**